output_chunks.cc

```diff
@@ -875,58 +875,52 @@ template <typename E>
 void DynsymSection<E>::sort_symbols(Context<E> &ctx) {
   Timer t(ctx, "sort_dynsyms");
 
-  struct T {
-    Symbol<E> *sym;
-    i32 idx;
-    u32 hash;
-
-    bool is_local() const {
-      return sym->esym().st_bind == STB_LOCAL;
-    }
-  };
-
-  std::vector<T> vec(symbols.size());
-
-  for (i32 i = 1; i < symbols.size(); i++)
-    vec[i] = {symbols[i], i, 0};
+  // The first entry of an ELF symbol table is a null symbol
+  const auto first_nonnull_symbol =
+      std::min(std::begin(symbols) + 1, std::end(symbols));
 
   // In any ELF file, local symbols should precede global symbols.
-  tbb::parallel_sort(vec.begin() + 1, vec.end(), [](const T &a, const T &b) {
-    return std::tuple(a.is_local(), a.idx) < std::tuple(b.is_local(), b.idx);
```

Comment:
Does this do what the comment says it does? Because `cmp( (1, idx) , (0, idx)` puts the local symbols _after_ the global symbols if I'm reading this correctly

Reply:
This piece of code looks odd in the first place, as the dynamic symbol table contains only global symbols. I don't think that putting any local symbol to .dynsym makes sense.

Reply:
Not familiar with linkers (reading MattPD's resources and learning as I work on this project), but this would explain why the existing code didn't exhibit any notable defects.
We could move the local symbol logic to an debug level assert, and assume we're only dealing with global symbols

Reply:
Hey, it looks like in the `i386-tls-gd` test case, `mold`@`fc5469` generates a local symbol in the executable `exe` -- this is not the case for the default clang linker
```
(base) gjye@gjye:~/clone/mold/test/tmp/i386-tls-gd$ readelf --dyn-syms exe

Symbol table '.dynsym' contains 7 entries:
   Num:    Value  Size Type    Bind   Vis      Ndx Name
     0: 00000000     0 NOTYPE  LOCAL  DEFAULT  UND 
     1: 00200570     4 OBJECT  GLOBAL DEFAULT  UND _IO_stdin_used
     2: 00000000     4 TLS     LOCAL  DEFAULT   20 bar
readelf: Warning: local symbol 2 found at index >= .dynsym's sh_info value of 1
     3: 00000000     0 TLS     GLOBAL DEFAULT  UND foo
     4: 00201010     0 FUNC    GLOBAL DEFAULT  UND printf@GLIBC_2.0 (3)
     5: 00201020     0 FUNC    GLOBAL DEFAULT  UND __libc_start_main@GLIBC_2.0 (3)
     6: 00201030     0 FUNC    GLOBAL DEFAULT  UND ___tls_get_addr@GLIBC_2.3 (2)
(base) gjye@gjye:~/clone/mold/test/tmp/i386-tls-gd$ readelf --dyn-syms exe-ld

Symbol table '.dynsym' contains 7 entries:
   Num:    Value  Size Type    Bind   Vis      Ndx Name
     0: 00000000     0 NOTYPE  LOCAL  DEFAULT  UND 
     1: 00000000     0 FUNC    GLOBAL DEFAULT  UND printf@GLIBC_2.0 (2)
     2: 00000000     0 FUNC    GLOBAL DEFAULT  UND ___tls_get_addr@GLIBC_2.3 (3)
     3: 00000000     0 NOTYPE  WEAK   DEFAULT  UND __gmon_start__
     4: 00000000     0 FUNC    GLOBAL DEFAULT  UND __libc_start_main@GLIBC_2.0 (2)
     5: 00000000     0 TLS     GLOBAL DEFAULT  UND foo
     6: 0804a004     4 OBJECT  GLOBAL DEFAULT   17 _IO_stdin_used